src/stdlib_linalg_inverse.fypp

```diff
@@ -0,0 +1,165 @@
+     use stdlib_linalg_constants
+     use stdlib_linalg_lapack, only: getri,getrf,stdlib_ilaenv
+     use stdlib_linalg_state, only: linalg_state_type, linalg_error_handling, LINALG_ERROR, &
+         LINALG_INTERNAL_ERROR, LINALG_VALUE_ERROR
+     implicit none(type,external)
+
+     character(*), parameter :: this = 'inverse'  
+
+     contains
+
+     elemental subroutine handle_getri_info(info,lda,n,err)
+         integer(ilp), intent(in) :: info,lda,n
+         type(linalg_state_type), intent(out) :: err
+
+         ! Process output
+         select case (info)
+            case (0)
+                ! Success
+            case (:-1)
+                err = linalg_state_type(this,LINALG_ERROR,'invalid matrix size a=',[lda,n])
+            case (1:)
+                ! Matrix is singular
+                err = linalg_state_type(this,LINALG_ERROR,'singular matrix')
+            case default
+                err = linalg_state_type(this,LINALG_INTERNAL_ERROR,'catastrophic error')
+         end select
+         
+     end subroutine handle_getri_info
+
+     #:for rk,rt,ri in RC_KINDS_TYPES
+     #:if rk!="xdp" 
+     ! Compute the in-place square matrix inverse of a
+     module subroutine stdlib_linalg_invert_inplace_${ri}$(a,pivot,err)
+         !> Input matrix a[n,n]. On return, A is destroyed and replaced by the inverse
+         ${rt}$, intent(inout) :: a(:,:)
+         !> [optional] Storage array for the diagonal pivot indices
+         integer(ilp), optional, intent(inout), target :: pivot(:)
+         !> [optional] state return flag. On error if not requested, the code will stop
+         type(linalg_state_type), optional, intent(out) :: err
+
+         !> Local variables
+         type(linalg_state_type) :: err0
+         integer(ilp) :: lda,n,info,nb,lwork,npiv
+         integer(ilp), pointer :: ipiv(:)
+         ${rt}$, allocatable :: work(:)
+
+         !> Problem sizes
+         lda  = size(a,1,kind=ilp)
+         n    = size(a,2,kind=ilp)
+
+         ! Has a pre-allocated pivots storage array been provided? 
+         if (present(pivot)) then 
+            ipiv => pivot
+         else
+            allocate(ipiv(n))
+         endif
+         npiv = size(ipiv,kind=ilp)
+
+         if (lda<1 .or. n<1 .or. lda/=n .or. npiv<n) then
+            err0 = linalg_state_type(this,LINALG_VALUE_ERROR,'invalid matrix size: a=',[lda,n], &
+                                                             ', pivot=',npiv)
+            if (.not.present(pivot)) deallocate(ipiv)
+            call linalg_error_handling(err0,err)
+            return
+         end if
+
+         ! Factorize matrix (overwrite result)
+         call getrf(lda,n,a,lda,ipiv,info)
+
+         ! Return codes from getrf and getri are identical
+         if (info==0) then
+
+            ! Get optimal worksize (returned in work(1)) (inflate by a 5% safety margin)
+            nb = stdlib_ilaenv(1,'${ri}$getri',' ',n,-1,-1,-1)
+            lwork = ceiling(1.05*n*nb,kind=ilp)
```

Comment:
We should be careful with integer overflow in this case (I have been bitten several times ;). It might be good to have a check for `lwork`

Reply:
[505f30a](https://github.com/fortran-lang/stdlib/pull/828/commits/505f30a2a38d571429439d36e51670be8df4baf2)